Add graceful-restart support

karampok opened the pull request:
/kind feature

**What this PR does / why we need it**:
Issue covers #2368 

**Special notes for your reviewer**:

**Release note**:


```release-note
TODO
```

karampok commented:
Note to discuss:
```
Graceful restart configuration changed, reset this peer to take effect
% The Graceful Restart command used is not valid at this moment.
line 76: Failure to communicate[13] to bgpd, line:  neighbor 11.11.11.1 graceful-restart

[311|bfdd] done
[294|zebra] done
[300|bgpd] Configuration file[/etc/frr/frr.conf] processing failure: 13
2024-04-26 13:18:54,515 WARNING: frr-reload.py failed due to
vtysh (exec file) exited with status 13
Failed to fully apply configuration file 1 seconds
Caught SIGHUP and acquired lock! Reloading FRR..
Checking the configuration file syntax
```
it seems it fails, and retries to apply that the speaker/frr-reloader.

karampok commented:
The above might be related to https://github.com/FRRouting/frr/issues/8403

karampok commented:
In my test: I achieved zero downtime if I have graceful restart either in global `bgp graceful restarst` or in neighbor mode `neighbor x.y.z.k graceful restart` and having that in  `bgp graceful-restart preserve-fw-state`. 

(Test= in runtime I delete the speaker and see if the client loses connectivity)

The second command that enables the F -bit is only available in global mode, not available per neighbor. Therefore a solution which we add graceful-restart in neighbor level and f-bit in global (because is not safe to just have that there by default) seem not optimal.

@fedepaol How much do you object to have it global (one for enable gracefull restart/one to on/off f-bit), and also have the option to opt out per neighbor? Any other ideas?

karampok commented:
https://github.com/FRRouting/frr/issues/15880 created this issue upstream, even though we are not really blocked on that (well we will get an error but doe not look like it has an impact)

fedepaol commented:
> The second command that enables the F -bit is only available in global mode, not available per neighbor. Therefore a solution which we add graceful-restart in neighbor level and f-bit in global (because is not safe to just have that there by default) seem not optimal.
> 
> @fedepaol How much do you object to have it global (one for enable gracefull restart/one to on/off f-bit), and also have the option to opt out per neighbor? Any other ideas?

Won't the global one affect only peers with gr enabled? If so, doesn't sound too terrible to always enable it.

karampok commented:
My preference would be to be configurable to be in the safe side. I cannot find argument against always having it (or being default on) but no idea for which case this option exists.

fedepaol commented on api/v1beta2/bgppeer_types.go:
```diff
@@ -85,6 +85,12 @@ type BGPPeerSpec struct {
+	// known routes while the routing protocol information is being restored.
+	// Needed for FRR mode only.
+	// +optional
+	GracefulRestart bool `json:"gracefulRestart,omitempty"`
```
nit: EnableGracefulRestart

karampok replied:
I can change that, sure

karampok replied:
done

fedepaol commented on api/v1beta2/bgppeer_types.go:
nit on the commit subject: this commit does more than adding the field to the api, we are adding the support end to end (which is totally fine, but let's change the subject to "add graceful restart support"

karampok replied:
I have three commits and the e2e is added on its own commit, no?

karampok replied:
done

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -427,6 +427,33 @@ var _ = ginkgo.Describe("BGP", func() {
+		err := ConfigUpdater.Update(resources)
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, c := range FRRContainers {
```
I think it'd make sense to have a real e2e test where we shut down a speaker and we check the traffic is still flowing.

First thing that comes to my mind is, focus on a node, exclude it from the daemonset selector, have a pod running only there with etp=local, ensure the traffic flows when the speaker is down

fedepaol replied:
It's probably easier if we raise the graceful restart timers of the peered neighbors, so we'll have control

fedepaol replied:
Let's start with the default and see if it's enough then

oribon commented on api/v1beta2/bgppeer_types.go:
```diff
@@ -85,6 +85,13 @@ type BGPPeerSpec struct {
 
+	// EnableGracefulRestart allows BGP peer to continue to forward data packets along
+	// known routes while the routing protocol information is being restored.
+	// Needed for FRR mode only.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="EnableGracefulRestart cannot be changed after creation, because GR requires rest BGP session"
```
nit: "Needed" should be more about "Supported" here?
also I'd put in the description the point about "this field is immutable because it requires restart of the BGP session" and keep the validation message "EnableGracefulRestart cannot be changed after creation"

karampok replied:
sounds good, done

oribon commented on internal/config/validation.go:
```diff
@@ -41,6 +41,9 @@ func DiscardFRROnly(c ClusterResources) error {
 			return fmt.Errorf("peer %s has connect time set on native bgp mode", p.Spec.Address)
 		}
+		if p.Spec.EnableGracefulRestart {
+			return fmt.Errorf("peer %s has enabled GR (GracefulRestart) flag set on native bgp mode", p.Spec.Address)
```
```suggestion
			return fmt.Errorf("peer %s has GracefulRestart flag set on native bgp mode", p.Spec.Address)
```

karampok commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +165,65 @@ var _ = ginkgo.Describe("BGP", func() {
+				})
+			}
+
+			err = metallb.RestartSpeakerPods(cancel, cs)
```
What do you mean "are restarted". Are you suggesting  blocking in a function which will check speakers are restarted and then continue to an eventuallly gingko body that loops over all external frr docker containers doing in serial (validateServiceNoWait)?

karampok replied:
refactored

karampok commented on e2etest/bgptests/bgp.go:
```diff
@@ -98,9 +101,25 @@ var _ = ginkgo.Describe("BGP", func() {
 	ginkgo.BeforeEach(func() {
+
+		clientconfig := k8sclient.RestConfig()
+		var err error
```
I had tod that because  otherwise it fails like https://github.com/metallb/metallb/actions/runs/9267228538/job/25493436433 (a wip commit that I reverse)

karampok replied:
There is some state that changes because I delete the speaker containers, so it must that logic go per test not per suite

fedepaol replied:
ah, I saw the reason. The FRR provider holds a reference to the speaker / frr-k8s pods.

This is a bit confusing though. FRRProvider is global, we initialize it two different times and we use it also in other files. I have the feeling this might bit us in the future (for example if we add a new context in a new file). 
How about giving the FRRProvider a Refresh() method that updates the speaker? So we re-load the proper speakers after any test that restarts the speakers?

Either that, or we fetch the speakers every time we call Execute. Shouldn't be terrible either as it fetches them from the local cache.

karampok replied:
I could go either, maybe the second seems simpler. Do you mean changing here https://github.com/metallb/metallb/blob/95274d94dffe0baed35330ebbbbe3f82282a19d1/e2etest/pkg/frr/provider/provider.go#L49 ? Can you expand?

karampok replied:
done

karampok commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +165,65 @@ var _ = ginkgo.Describe("BGP", func() {
+			eg, ctx := errgroup.WithContext(ctx)
+			for _, c := range FRRContainers {
+				validateService(svc, allNodes.Items, c)
+				cc := c // https://go.dev/doc/faq#closures_and_goroutines
```
done (probably not needed anymore though or will be done)

karampok commented on e2etest/pkg/metallb/metallb.go:
```diff
@@ -32,6 +32,38 @@ func init() {
+
+		ret := true
+		for _, p := range pods {
+			ret = ret && k8s.PodIsReady(p)
```
done

karampok commented on e2etest/pkg/metallb/metallb.go:
```diff
@@ -32,6 +32,38 @@ func init() {
+	}
+
+	f := func(context.Context) (bool, error) {
+		pods, err := SpeakerPods(cs)
```
TODO(i time.Sleep(5) for the time being)

karampok replied:
done

karampok commented on e2etest/pkg/metallb/metallb.go:
```diff
@@ -32,6 +32,38 @@ func init() {
 	}
 }
 
+func RestartSpeakerPods(cancel context.CancelFunc, cs clientset.Interface) error {
```
made it non blocksing as we discussed

karampok commented on e2etest/pkg/metallb/metallb.go:
```diff
@@ -32,6 +32,38 @@ func init() {
+		}
+	}
+
+	f := func(context.Context) (bool, error) {
```
done, but is there a drawback to have it assigned to variable?

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +164,134 @@ var _ = ginkgo.Describe("BGP", func() {
+			Expect(metallb.RestartSpeakerPodsNoWait(cs)).NotTo(HaveOccurred())
+
+			Eventually(func() error {
+				c := func() error {
```
why do you need to assign to a variable? Can't you just iterate over the containers and assert from there?

karampok replied:
I find easier to read but not strong opinion, I will change it

karampok replied:
done

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +164,134 @@ var _ = ginkgo.Describe("BGP", func() {
+				}
+
+				if err := c(); err != nil {
+					if !errors.Is(err, ErrStaleRoute) {
```
why is staleroute skipped?

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +164,134 @@ var _ = ginkgo.Describe("BGP", func() {
+			//	}
+			// }()
+
+			Expect(metallb.RestartSpeakerPodsNoWait(cs)).NotTo(HaveOccurred())
```
I'd stick with the err := and the assert on the next line pattern.

karampok replied:
done

karampok commented:
@fedepaol I have completed the last commit with adding the e2e test (plus the reverse to see a failure without graceful restart). I think I addressed all your comments. Let me know if I miss something. Thanks

fedepaol commented on e2etest/pkg/metallb/metallb.go:
```diff
@@ -32,6 +33,35 @@ func init() {
 	}
 }
 
+func RestartSpeakerPodsNoWait(cs clientset.Interface) error {
```
Let's rename this, as it waits :)

karampok replied:
done

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +147,115 @@ var _ = ginkgo.Describe("BGP", func() {
+
+		ginkgo.Context("and when GR enabled", func() {
+			ginkgo.BeforeEach(func() {
+				grEnabled = true
```
let's have this a member of the assertDuringSpeakerRestart function (shouldFail, for example)

karampok replied:
not done, can you expand what you mean?

fedepaol replied:
instead of having a global that share some state, I'd make assertDuringSpeakerRestart pure with that regard and allow it to accept an extra parameter which is that bool. The same bool can then be a parameter of the table tests (bonus point, use a constant with the proper name instead of a bool).

karampok replied:
Currently I have two tables, exactly the same with the diff the beforeeach section. I think the way that the Describe scope have a "global" state, and that in the before each you define it (as you do with the frr config. gr to true) is more readable to understand the test behavior and fits better ginkgo approach. The description looks like 

```
BGP GracefulRestart, when speakers restart and when GR enabled dataplane should keep working (entry x,y,z)
BGP GracefulRestart, when speakers restart and when GR disabled dataplane should have a downtime (entry x,y,z)
```

If I add as extra to the function, then you suggest to actually merge the test into single context/table (right?) and have the  bool name hint  what we test inside the entry  args instead of the ginkgo description. tbh I would prefer to keep it that way but I can change it if you want.

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +147,115 @@ var _ = ginkgo.Describe("BGP", func() {
+					for _, c := range FRRContainers {
+						err := validateServiceNoWait(svc, allNodes.Items, c)
+						if err != nil {
+							if errors.Is(err, ErrStaleRoute) {
```
let's avoid extra nesting and assert stale route first

karampok replied:
done

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +147,115 @@ var _ = ginkgo.Describe("BGP", func() {
+				for _, c := range FRRContainers {
+					err := validateServiceNoWait(svc, allNodes.Items, c)
+					if err != nil {
+						if errors.Is(err, ErrStaleRoute) {
```
same

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +147,115 @@ var _ = ginkgo.Describe("BGP", func() {
+						if errors.Is(err, ErrStaleRoute) {
+							continue // when GR, is normal to observe stale routes
+						}
+						Expect(err).NotTo(HaveOccurred(), "downtime was observed")
```
this can be outside of the if

karampok replied:
done

karampok commented:
Another version pushed that does make the function not depend on external variable

fedepaol commented on e2etest/bgptests/bgp.go:
```diff
@@ -147,6 +147,117 @@ var _ = ginkgo.Describe("BGP", func() {
+		ginkgo.Context("and when GR enabled", func() {
+
+			assertDuringSpeakerRestartWithGR := func(pairingIPFamily ipfamily.Family, poolAddresses []string, tweak testservice.Tweak) {
+				assertDuringSpeakerRestart(true, pairingIPFamily, poolAddresses, tweak)
```
nit: let's define a constant "EnableGracefulRestart = true" and let's use that instead of true / false (also when checking)

karampok replied:
done

fedepaol commented on e2etest/pkg/wget/wget.go:
```diff
@@ -25,7 +25,7 @@ func Do(address string, exc executor.Executor) error {
 	// Retry loop to handle wget NetworkFailure errors
 	for {
-		out, err = exc.Exec("wget", "-O-", "-q", address, "-T", "60")
+		out, err = exc.Exec("wget", "-O-", "-q", address, "-T", "5")
```
is this change related?

fedepaol commented on e2etest/pkg/frr/provider/provider.go:
```diff
@@ -43,22 +47,28 @@ func NewFRRMode(r *rest.Config) (Provider, error) {
+	_, ok := speakers[name]
 	if !ok {
-		return nil, fmt.Errorf("speakers %s/%s not found in known speakers %v", ns, name, f.speakers)
+		return nil, fmt.Errorf("speakers %s/%s not found in known speakers %v", ns, name, speakers)
```
nit: speaker %s not found in..

fedepaol commented on e2etest/pkg/jigservice/jig.go:
```diff
@@ -415,7 +415,7 @@ func (j *TestJig) UpdateService(ctx context.Context, update func(*v1.Service)) (
 // WaitForLoadBalancer waits the given service to have a LoadBalancer, or returns an error after the given timeout
 func (j *TestJig) WaitForLoadBalancer(ctx context.Context, timeout time.Duration) (*v1.Service, error) {
-	ginkgo.GinkgoWriter.Printf("Waiting up to %v for service %q to have a LoadBalancer", timeout, j.Name)
+	ginkgo.GinkgoWriter.Printf("\tWaiting up to %v for service %q to have a LoadBalancer\n", timeout, j.Name)
```
is this related? If not but makes sense, please add it in a separte commit

fedepaol commented on e2etest/pkg/metallb/metallb.go:
```diff
@@ -32,6 +33,35 @@ func init() {
+		}
+		return true, nil
+	})
+
```
nit: unnecessary line

fedepaol commented:
few nits, almost there!

karampok reviewed (commented):
What about now :) let me know if you want to rebase. Thanks

fedepaol commented:
> What about now :) let me know if you want to rebase. Thanks

rebase is not necessary as we use merge queues.
LGTM, sending to merge queue, thanks!
